File: edukate-backend/src/main/java/io/github/sanyavertolet/edukate/backend/entities/files/FileObject.java
package io.github.sanyavertolet.edukate.backend.entities.files;

import io.github.sanyavertolet.edukate.storage.keys.FileKey;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "file_objects")
public class FileObject {
    @Id
    private String id;

    @Indexed(unique = true)
    private String keyPath;

    private FileKey key;

    private String type;

    private String ownerUserId;

    private FileObjectMetadata metadata;

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;

    @Builder.Default
    private Integer metaVersion = 1;
}

File: edukate-backend/src/main/kotlin/io/github/sanyavertolet/edukate/backend/entities/files/FileObject.kt
package io.github.sanyavertolet.edukate.backend.entities.files

import io.github.sanyavertolet.edukate.storage.keys.FileKey
import java.time.Instant
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.Id
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.relational.core.mapping.Table

@Table("file_objects")
data class FileObject(
    @Id val id: Long? = null,
    val keyPath: String,
    val key: FileKey,
    val type: String,
    val ownerUserId: Long,
    val metadata: FileObjectMetadata,
    @CreatedDate val createdAt: Instant? = null,
    @LastModifiedDate val updatedAt: Instant? = null,
    val metaVersion: Int = 1,
) {
    fun withStorageState(
        keyPath: String,
        key: FileKey,
        type: String,
        ownerUserId: Long,
        metadata: FileObjectMetadata,
    ): FileObject = copy(keyPath = keyPath, key = key, type = type, ownerUserId = ownerUserId, metadata = metadata)

    companion object {
        @JvmStatic
        fun fromStorageState(keyPath: String, key: FileKey, type: String, ownerUserId: Long, metadata: FileObjectMetadata) =
            FileObject(keyPath = keyPath, key = key, type = type, ownerUserId = ownerUserId, metadata = metadata)
    }
}
